src/protocol/attribution/apply_attribution_window.rs

```diff
@@ -0,0 +1,351 @@
+
+            let previous_stop_bit = &stop_bits[index - step_size];
+            let current_stop_bit = &stop_bits[index];
+            let previous_t_delta = &t_delta[index - step_size];
+
+            futures.push(async move {
+                try_join(
+                    // if the current stop_bit is 0, don't accumulate
+                    c1.multiply(record_id, previous_t_delta, current_stop_bit),
+                    // next `stop_bit`
+                    c2.multiply(record_id, previous_stop_bit, current_stop_bit),
+                )
+                .await
+            });
+        }
+
+        let results = try_join_all(futures).await?;
+
+        results
+            .into_iter()
+            .enumerate()
+            .for_each(|(i, (delta, stop_bit))| {
+                t_delta[i + step_size] += &delta;
+                stop_bits[i + step_size] = stop_bit;
+            });
+    }
+
+    Ok(t_delta)
+}
+
+/// Creates a vector of trigger values where values are set to `0` if the time delta
+/// from their nearest source event exceed the specified attribution window cap.
+///
+/// This protocol executes the bit-decomposition protocol in order to compare shares
+/// of time deltas in `F`.
+///
+/// # Errors
+/// Fails if the bit-decomposition, bitwise comparison, or multiplication fails.
+async fn zero_out_expired_trigger_values<F, C, T>(
+    ctx: &C,
+    input: &[MCApplyAttributionWindowInputRow<F, T>],
+    time_delta: &[T],
+    cap: u32,
+) -> Result<Vec<T>, Error>
+where
+    F: Field,
+    C: Context<F, Share = T>,
+    T: Arithmetic<F>,
+{
+    // Compare the accumulated timestamp deltas with the specified attribution window
+    // cap value, and zero-out trigger event values that exceed the cap.
+    let c = ctx.clone().set_total_records(input.len());
+    let bit_decomposition_ctx = c.narrow(&Step::TimeDeltaBitDecomposition);
+    let cmp_ctx = c.narrow(&Step::TimeDeltaLessThanCap);
+    let mul_ctx = c.narrow(&Step::CompareBitTimesTriggerValue);
+
+    let random_bits_generator =
+        RandomBitsGenerator::new(ctx.narrow(&Step::RandomBitsForBitDecomposition));
+    let rbg = &random_bits_generator;
+
+    try_join_all(
+        zip(input, time_delta)
+            .zip(repeat(ctx.share_known_value(F::ONE)))
+            .enumerate()
+            .map(|(i, ((row, delta), one))| {
+                let c1 = bit_decomposition_ctx.clone();
+                let c2 = cmp_ctx.clone();
+                let c3 = mul_ctx.clone();
+                let record_id = RecordId::from(i);
+
+                async move {
+                    let delta_bits = BitDecomposition::execute(c1, record_id, rbg, delta).await?;
+                    let compare_bit = one
+                        - &bitwise_greater_than_constant(c2, record_id, &delta_bits, cap.into())
+                            .await?;
+                    c3.multiply(record_id, &row.trigger_value, &compare_bit)
+                        .await
```

Comment:
Seems about right! 

On the one hand, I'm glad we have all of these powerful and useful protocols available for use, so that we can implement new functionality like this really quickly!

On the other hand, I'm crying inside because these are such expensive protocols!